### PR TITLE
new method without antithetic estimator, add gaussian likelihood with…

### PR DESCRIPTION
**Pull request by XiaRui1996**
1. New method without using antithetic estimator

- Speed up using $DW_i = g_i\hat{a}^T_{i-1}$, where $DW_i$ is gradient of parameter of the ith layer, $g_i$ is gradient w.r.t output of ith layer and $\hat{a}_i$ is input to ith layer, or output of (i-1)th layer. This is realized by `save_layer_input` and `save_layer_grad_output` and `Qg` for each module. (`Qv` will be the usual product, `Qg` will refer to speed up for small batch size. _TODO: for some module this should be do nothing_)
- Get rid of `plus_model` and `minus_model`, and store gradient info into `group['grad']` for each layer

2. Add `_prepare_input` to deal with inputs wrapped in a dictionary, this is the case especially for language datasets.
3. Add a Gaussian likelihood for which sigma is learnable and also updated through FishLeg, having auxiliary parameter etc.

### CODE REVIEW

**Comment by jamie-mcg** [comment]
Hi @XiaRui1996, please can you rebase this branch with main? i.e. run `git rebase main` from the command line inside this branch. Then push it again.

Thanks!

**Comment by XiaRui1996** [comment]
I have done `git rebase main` inside this branch and it says Current branch review/rui is up to date?

**Comment by jamie-mcg** [comment]
Ah OK, that's weird. Did you `git pull` the latest main branch first?

**Comment by XiaRui1996** [comment]
yeah, of course. But I doubt maybe not latest enough. Which part is missing?

**Comment by jamie-mcg** [comment]
Just sorted it! The branch was showing it was 1 commit behind main but it was just the license so I have just merged it in.

File: src/optim/FishLeg/fishleg_likelihood.py [hunk 1]
```diff
@@ -61,6 +66,12 @@ def draw(self, preds, **kwargs):
         """
         raise NotImplementedError
 
+    def get_parameters(self) -> List:
```
Comment (jamie-mcg) [inline src/optim/FishLeg/fishleg_likelihood.py:69]: Is this an abstract method? If we are requiring the list of learnable parameters to be defined (even if there may not be any) then we should probably decorate this and follow the `raise NotImplementedError`.

Basically, if someone defines a new likelihood and forgets to do this then it might be a hidden error if we don't through a `NotImplementedError`.

If it is non-essential then this is OK - I would just rewrite it as returning None.

Comment (XiaRui1996) [reply src/optim/FishLeg/fishleg_likelihood.py:69]: Because in the fishleg main optimiser, I will check whether the length of this function is > 0. So I return [] as default in here. If someone defines a new likelihood with parameters, they should return the parameters as a list in here.

Comment (jamie-mcg) [reply src/optim/FishLeg/fishleg_likelihood.py:69]: OK cool, so in that case, I think we should have it as a `raise NotImplementedError` in the FishLiklihood abstract class. Then define it explicitly for each likelihood below.

Since otherwise this may be missed in defining new likelihood classes.

File: src/optim/FishLeg/fishleg.py [hunk 2]
```diff
@@ -105,40 +112,35 @@ def __init__(
...
-        self.plus_model = copy.deepcopy(self.model)
-        self.minus_model = copy.deepcopy(self.model)
         self.sgd_lr = sgd_lr
-        self.lr = lr
+        self.fish_lr = fish_lr
         self.device = device
+        self.initialization = initialization
 
         self.model = self.init_model_aux(model).to(device)
+        self.likelihood = likelihood
```
Comment (jamie-mcg) [inline src/optim/FishLeg/fishleg.py:138]: I think if we are passing in the likelihood anyway, we can probably deprecate the need for `draw` and `nll` to be loaded in?

In the first instance, these can probably just be copied in as methods inside the `FishLeg` class. What do you think?

We can always allow for these methods to be overwritten if needed, but so far I haven't seen any of them look any different.

File: src/optim/FishLeg/fishleg.py [hunk 3]
```diff
@@ -105,40 +112,35 @@ def __init__(
...
+        initialization: str = 'uniform',
         device: str = "cpu",
+        num_steps = None
     ) -> None:
         self.model = model
-        self.plus_model = copy.deepcopy(self.model)
-        self.minus_model = copy.deepcopy(self.model)
         self.sgd_lr = sgd_lr
-        self.lr = lr
+        self.fish_lr = fish_lr
```
Comment (jamie-mcg) [inline src/optim/FishLeg/fishleg.py:133]: I think it would be best to stick to the hyperparameter names set out in the paper. But happy to be overruled on this.

File: src/optim/FishLeg/fishleg.py [hunk 4]
```diff
@@ -179,6 +228,7 @@ def __init__(
         self.beta = beta
         self.pre_aux_training = pre_aux_training
         self.step_t = 0
+        self.store_g = True
```
Comment (jamie-mcg) [inline src/optim/FishLeg/fishleg.py:231]: Is this changeable? If so, it belongs in the init with True as a default. If not then probably not needed here.

File: src/optim/FishLeg/fishleg.py [hunk 5]
```diff
@@ -226,53 +293,85 @@ def update_aux(self) -> None:
...
+                group['grad'][i].copy_(grad) 
+        
+        g_norm = torch.sqrt(g2) 
+
+        self.zero_grad()
+        # How to better implement this?
+        # The hook is not updated here, locally, only the gradient to the parameters g.grad is updated
+        self.store_g = False
+        self.nll(self.model, samples).backward()
+        self.store_g = True
```
Comment (jamie-mcg) [inline src/optim/FishLeg/fishleg.py:322]: What is this doing here? Does this store_g parameter affect the model somehow?

Comment (XiaRui1996) [reply src/optim/FishLeg/fishleg.py:322]: The store_g need to be True for the forward storage of input (activated inside `.nll`) and backward storage of gradient (activated by `.backward()`) to be executed. I don't want to execute them, because `Qg()` will still use gradient outside the aux_update step. This backward step will only update the gradient attach to parameter.grad, but not layer._a and layer._g.

File: src/optim/FishLeg/fishleg.py (hunk 4 above)
Comment (XiaRui1996) [reply src/optim/FishLeg/fishleg.py:231]: Yes, changeable and initialised as True.

File: src/optim/FishLeg/fishleg.py (hunk 2 above)
Comment (XiaRui1996) [reply src/optim/FishLeg/fishleg.py:138]: Yep, I think we can rewrite this better by setting default `draw` and `nll`. For language model, they do look a bit different. An example below. 

```
def nll(model, inputs):
    outputs = model(**inputs, head='lm_head')
    pred_y = outputs['logits'][..., :-1, :].contiguous()
    data_y = inputs['labels'][..., 1:].contiguous()
    return -Categorical(
                        logits=pred_y.view(-1, model.config.vocab_size)
                    ).log_prob(data_y.view(-1)).mean()
```

File: src/optim/FishLeg/fishleg_likelihood.py [hunk 6]
```diff
@@ -107,18 +158,11 @@ def __init__(self, device: str = "cpu") -> None:
...
-        result = (
-            -torch.sum(
-                torch.log(1 - preds + 1e-5) * (1.0 - observations)
-                + torch.log(preds + 1e-5) * observations
-            )
-            / preds.shape[0]
-        )
-        return result
+        bce = torch.sum(preds * (1. - observations) + torch.nn.Softplus()(-preds))
+        return bce / preds.shape[0]
```
Comment (jamie-mcg) [inline src/optim/FishLeg/fishleg_likelihood.py:162]: Please can we check this refactor of nll? Further to our conversation on Slack, this may be causing an issue?

Also the method below where probs is substituted for logits.

File: src/optim/FishLeg/fishleg.py [hunk 7]
```diff
@@ -226,74 +393,113 @@ def update_aux(self) -> None:
...
+
+        if True:
             g2 = 0.0
             for group in self.param_groups:
                 for p in group["params"]:
-                    g2 += p.grad.data.norm(p=2) ** 2
-            g2 = torch.sqrt(g2)
+                    grad = p.grad.data
+                    g2 = g2 + torch.sum(grad * grad)
+            g_norm = torch.sqrt(g2)
```
Comment (jamie-mcg) [inline src/optim/FishLeg/fishleg.py:410]: Doesn't look like we are using this anymore? Is this correct?

File: src/optim/FishLeg/fishleg.py [hunk 8]
```diff
@@ -226,74 +393,113 @@ def update_aux(self) -> None:
...
-                    p.data + d_p * self.eps
-                )
-                self.minus_model._modules[name]._parameters[para_name] = (
-                    p.data - d_p * self.eps
-                )
+            for p, g, d_p in zip(group["params"], group["grad"], qg):
 
-                aux_loss -= 2 * torch.sum(g * d_p)
-                aux_loss += self.damping * d_p.norm(p=2) ** 2
+                grad = p.grad.data
```
Comment (jamie-mcg) [inline src/optim/FishLeg/fishleg.py:424]: This gradient is not normalised - is this correct?

File: src/optim/FishLeg/fishleg.py [hunk 9]
```diff
@@ -226,74 +393,113 @@ def update_aux(self) -> None:
...
 
-        h_plus = self.nll(self.plus_model, data_x, pred)
-        h_minus = self.nll(self.minus_model, data_x, pred)
+        quad_term = quad_term**2
 
-        aux_loss += (h_plus + h_minus) / (self.eps**2)
-        aux_loss.backward()
+        aux_loss = 0.5 * (reg_term + quad_term) - linear_term
+        if self.normalization:
+            aux_loss = aux_loss / g2
```
Comment (jamie-mcg) [inline src/optim/FishLeg/fishleg.py:433]: We are back to normalising the loss, is there a reason we are doing this?

File: src/optim/FishLeg/fishleg.py [hunk 10]
```diff
@@ -226,74 +393,113 @@ def update_aux(self) -> None:
...
+                linear_term = linear_term + torch.sum(g * d_p)
+                reg_term = reg_term + self.damping * torch.sum(d_p * d_p)
 
-        h_plus = self.nll(self.plus_model, data_x, pred)
-        h_minus = self.nll(self.minus_model, data_x, pred)
+        quad_term = quad_term**2
 
-        aux_loss += (h_plus + h_minus) / (self.eps**2)
-        aux_loss.backward()
+        aux_loss = 0.5 * (reg_term + quad_term) - linear_term
```
Comment (jamie-mcg) [inline src/optim/FishLeg/fishleg.py:431]: When I run this in a training run, I see a negative loss every now and then. Is this what you also see?

**Review by jamie-mcg — COMMENTED** [review]
I'm seeing some weird characteristics when training on MNIST - have you tried this already?

File: src/optim/FishLeg/fishleg.py (hunk 10 above)
Comment (XiaRui1996) [reply src/optim/FishLeg/fishleg.py:431]: In my experiment of fine-tuning yes, if the model is pre-trained to a local minimum, loss is negative.

Comment (jamie-mcg) [reply src/optim/FishLeg/fishleg.py:431]: OK, so what is the process for training from scratch? In this case I am still following the .step() update and the aux loss seems to be negative a lot during a training run.

File: src/optim/FishLeg/fishleg.py [hunk 11]
```diff
@@ -105,82 +112,92 @@ def __init__(
...
         weight_decay: float = 1e-5,
         beta: float = 0.9,
-        update_aux_every: int = -3,
-        aux_lr: float = 1e-3,
+        update_aux_every: int = 10,
+        aux_lr: float = 1e-4,
         aux_betas: Tuple[float, float] = (0.9, 0.999),
         aux_eps: float = 1e-8,
         damping: float = 1e-5,
         pre_aux_training: int = 10,
```
Comment (jamie-mcg) [inline src/optim/FishLeg/fishleg.py:124]: This is also no longer used.

File: src/optim/FishLeg/fishleg.py (hunk 10 above)
Comment (ghennequin) [reply src/optim/FishLeg/fishleg.py:431]: (the average over `g` is `-0.5 Trace(F^{-1} cov(g))`)

Comment (XiaRui1996) [reply src/optim/FishLeg/fishleg.py:431]: Will damping be a big effect here? as `Qg` is actually `(F + \gamma I)^{-1}g`
